consensus/src/processes/coinbase.rs

```diff
@@ -64,62 +105,121 @@ impl CoinbaseManager {
+        // Collect all rewards from mergeset reds ∩ DAA window and create a
+        // single output rewarding all to the current block (the "merging" block)
+        let mut red_reward = 0u64;
+        for red in ghostdag_data.mergeset_reds.iter().filter(|h| !mergeset_non_daa.contains(h)) {
+            let reward_data = mergeset_rewards.get(red).unwrap();
+            red_reward += reward_data.subsidy + reward_data.total_fees;
+        }
+        if red_reward > 0 {
+            outputs.push(TransactionOutput::new(red_reward, miner_data.script_public_key.clone()));
+        }
+
+        // Build the current block's payload
+        let subsidy = self.calc_block_subsidy(daa_score);
+        let payload = self.serialize_coinbase_payload(&CoinbaseData { blue_score: ghostdag_data.blue_score, subsidy, miner_data })?;
+
+        Ok(CoinbaseTransactionTemplate {
+            tx: Transaction::new(constants::TX_VERSION, vec![], outputs, 0, subnets::SUBNETWORK_ID_COINBASE, 0, payload),
+            has_red_reward: red_reward > 0,
+        })
+    }
+
+    pub fn serialize_coinbase_payload(&self, data: &CoinbaseData) -> CoinbaseResult<Vec<u8>> {
+        let script_pub_key_len = data.miner_data.script_public_key.script().len();
+        if script_pub_key_len > self.coinbase_payload_script_public_key_max_len as usize {
+            return Err(CoinbaseError::PayloadScriptPublicKeyLenAboveMax(
+                script_pub_key_len,
+                self.coinbase_payload_script_public_key_max_len,
+            ));
+        }
+        let payload: Vec<u8> = data.blue_score.to_le_bytes().iter().copied()                    // Blue score                   (u64)
+            .chain(data.subsidy.to_le_bytes().iter().copied())                                  // Subsidy                      (u64)
+            .chain(data.miner_data.script_public_key.version().to_le_bytes().iter().copied())   // Script public key version    (u16)
+            .chain((script_pub_key_len as u8).to_le_bytes().iter().copied())                    // Script public key length     (u8)
+            .chain(data.miner_data.script_public_key.script().iter().copied())                  // Script public key            
+            .chain(data.miner_data.extra_data.iter().copied())                                  // Extra data
+            .collect();
+
+        Ok(payload)
+    }
+
+    pub fn modify_coinbase_payload(&self, mut payload: Vec<u8>, miner_data: &MinerData) -> CoinbaseResult<Vec<u8>> {
+        let script_pub_key_len = miner_data.script_public_key.script().len();
+        if script_pub_key_len > self.coinbase_payload_script_public_key_max_len as usize {
+            return Err(CoinbaseError::PayloadScriptPublicKeyLenAboveMax(
+                script_pub_key_len,
+                self.coinbase_payload_script_public_key_max_len,
+            ));
+        }
+
+        payload.truncate(LENGTH_OF_BLUE_SCORE + LENGTH_OF_SUBSIDY); // Keep only blue score and subsidy
+        payload.extend(
```

Comment:
nit: In go we have an optimization where we create a new vector only if the length was changed (which almost never happens)

Reply:
vec.truncate keeps capacity as is, so this is equal